### ISSUE TITLE
Adding bindings to non-existent shape fails silently

### ISSUE THREAD

**Issue by trygve-aaberge-adsk** [tldraw/tldraw#4558]
### What happened?

If you try to add a binding through `store.put` and supply a shapeId that doesn't exist in `fromId` or `toId` the call to put just adds a binding that is broken. Even if a shape with that id is added later, the binding is still broken unless it's added again after the shape was added.

Previously supplying a `fromId` that didn't exist would result in a `TypeError: Cannot read properties of undefined (reading 'id')`, but after #4506 it's just ignored. The TypeError isn't great either of course, but I think throwing a descriptive error in this case would be good. Probably a bit earlier in the stack though, before the binding is created. And for both `fromId` and `toId`. Alternatively, at least print an error to the console so you don't miss it completely.

The reason we had this happening is that we sync between clients with Yjs and use YKeyValue which may optimize a sequence of adds and updates to just one add (e.g. if the client is disconnected, and several changes are done before they connect again). This changed the order of the changes, and we applied each change sequentially with a separate `store.put` call for each change. While investigating this I realized that if we apply all changes in the same `store.put` call, it works fine, so in that sense the problem is solved. However, I'd still like handling of it in case something similar happens again.

### How can we reproduce the bug?

Create a binding with a `fromId` or `toId` that doesn't point to an existing shape and put it in the store.

I've created an example here: https://codesandbox.io/p/sandbox/tldraw-example-forked-f422rx

### What browsers are you seeing the problem on?

Chrome

### Where did this happen?

The developer package

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by MitjaBezensek** [comment]
Hey! Thanks for bringing this up. I definitely agree that we should do a better job in communicating failures in our api calls. A while back I opened up an RFC for this, would love to hear your take on it. Here's the RFC:
https://github.com/tldraw/tldraw/issues/3637

That said I think that the mentioned PR (the one that you linked) at least for now does what we want it to do. We don't want to allow creating bindings to non existing shapes since this will cause issues down the line. We want to detect these kinds of issues early instead of them propagating to the editor and something blowing up much later.

**Comment by trygve-aaberge-adsk** [comment]
Thanks, I'll take a look at the RFC.

> That said I think that the mentioned PR (the one that you linked) at least for now does what we want it to do. We don't want to allow creating bindings to non existing shapes since this will cause issues down the line. We want to detect these kinds of issues early instead of them propagating to the editor and something blowing up much later.

I agree not allowing you to create the binding would be good, but it doesn't look like that's what happening with that PR? Doesn't it only return early in a handler? If you try to create a binding as in the codesandbox example I provided, you'll see that it's still created even though it points to shapes that don't exist.

**Comment by MitjaBezensek** [comment]
> I agree not allowing you to create the binding would be good, but it doesn't look like that's what happening with that PR? Doesn't it only return early in a handler? If you try to create a binding as in the codesandbox example I provided, you'll see that it's still created even though it points to shapes that don't exist.

Oh, you are using store.put, I was thinking of the editor APIs, like the `createBindings` method, which skips bindings to non-existent shapes.
https://github.com/tldraw/tldraw/blob/main/packages/editor/src/lib/editor/Editor.ts#L5510

Maybe you could run all your actions inside `editor.run` and call the editor apis instead of going straight to the store?

**Comment by trygve-aaberge-adsk** [comment]
> Maybe you could run all your actions inside editor.run and call the editor apis instead of going straight to the store?

Then you're basically saying `store.put` shouldn't be used?

The example I posted is simplified. In reality I'm doing this when syncing from a remote store and other clients, so it runs completely outside of React. So I don't think I have any way to access `editor` there, and I also would think it makes more sense to put those directly in the store than going via editor APIs?

**Comment by MitjaBezensek** [comment]
> Then you're basically saying `store.put` shouldn't be used?
>  
> The example I posted is simplified. In reality I'm doing this when syncing from a remote store and other clients, so it runs completely outside of React. So I don't think I have any way to access `editor` there, and I also would think it makes more sense to put those directly in the store than going via editor APIs?

Looks like you are already going through the editor apis - that's where the origin of changes is. Which means the the logic for not allowing bindings to non-existent shapes already checked that everything was fine when creating the binding. So the store is in nice and consistent state. The issue then occurs with the reordering of steps. Don't think we can make all the same checks happen again in the store itself. Also, in your case that's not what you'd want: somebody created a valid binding, so we should not ignore it. Feels like the solution to do all the changes in a single put might be the best way to go. Another option would be to enforce the correct ordering in your optimization step.

**Comment by trygve-aaberge-adsk** [comment]
> Looks like you are already going through the editor apis - that's where the origin of changes is. Which means that the logic for not allowing bindings to non-existent shapes already checked that everything was fine when creating the binding. So the store is in nice and consistent state. The issue then occurs with the reordering of steps.

Right

> Don't think we can make all the same checks happen again in the store itself. Also, in your case that's not what you'd want: somebody created a valid binding, so we should not ignore it. Feels like the solution to do all the changes in a single put might be the best way to go. Another option would be to enforce the correct ordering in your optimization step.

Yeah, for actually fixing the issue I agree we have to do it in a single put or make sure the order stays correct. I'm just saying that by silently ignoring the problem, it makes it harder to detect when we unintentionally have issues like this. That's why I would at least want a console.error when something unexpected happens, so we can easier notice it and fix the issue.

The reason we make individual put calls for each record appears to be that we have had issues where there was one bad record, which made the whole put with all the records fail if we did it in one call. Maybe `.put` could instead of throwing, insert all the valid records, and return which records were inserted and which failed?

I'm not sure how we ended up with a bad record. I'll ask my team mate which has worked on this for longer if he can provide some more context.

**Comment by ds300** [comment]
I can see why it's unexpected to fail silently here. I reckon we could add a suppressible warning. I'm hesitant to be throwing errors because according to our sentry issues, it seems like we do accidentally create broken bindings sometimes and we're not sure why or if it's even user-impacting. I'm also thinking ahead to situations where authz could prevent one shape involved in a binding from being available, but not the binding itself (this could have implications for how we delete bindings after one of the shapes is deleted :thinking:).

Related, we have this internal method `store.atomic(() => { ...do puts and deletes in here })` which will defer the `onAfter*` callbacks until the end of the block, and in this case it should result in the bindings index being updated properly, so if we made that public you can use it instead of having to .put everything in one go. Let me know if you want this and I'll look into it.

But maybe the real issue here is that the bindings index should be able to track partial/broken bindings in case they become whole again. I don't see a problem with that. I'll add it to my to-do list.

**Comment by ds300** [comment]
> authz as in authorization? Do you have cases where the user doesn't have access to all shapes?

yes authorization. We don't have per-shape permissions (yet) but our customers do. At the moment the shapes are only hidden during render, but the data is still sent across the wire to users who aren't allowed to see them. I think at some point we might want to allow avoiding that, but it's not a priority.
